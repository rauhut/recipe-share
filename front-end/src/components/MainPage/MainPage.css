.recipe-card-grid {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(340px, 1fr));
  grid-gap: 5%;
  padding: 20px;
  align-items: center;
  justify-content: center;
}
