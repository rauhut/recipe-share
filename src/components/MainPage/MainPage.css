h6 > a,
h6 > a:hover {
  color: black;
}
