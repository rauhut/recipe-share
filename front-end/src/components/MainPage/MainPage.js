import React from "react";
import {
  Card,
  CardImg,
  CardTitle,
  CardText,
  CardSubtitle,
  CardBody,
  Modal,
  Spinner,
} from "reactstrap";
import { Link } from "react-router-dom";
import "./MainPage.css";
import AddRecipeModal from "../AddRecipeModal/AddRecipeModal";

class MainPage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      recipes: [],
      isCreateRecipeOpen: false,
      isLoading: true,
    };
  }

  componentDidMount() {
    fetch("http://localhost:3000/recipes", {
      method: "get",
      headers: {
        "Content-Type": "application/json",
      },
    })
      .then((response) => {
        if (response.ok) {
          return response.json();
        } else {
          throw new Error("Something went wrong when trying to get recipes");
        }
      })
      .then((recipes) => {
        this.setState({ recipes: recipes.data });
        this.setState({ isLoading: false });
      })
      .catch((error) => console.log(error));
  }

  toggle = () => {
    this.setState({ isCreateRecipeOpen: !this.state.isCreateRecipeOpen });
  };

  navToRecipe = (id) => {
    this.props.history.push(`/recipe/${id}`);
  };

  render() {
    const { recipes, isCreateRecipeOpen } = this.state;
    return (
      <div>
        <h1>Recipe Share</h1>
        {this.state.isLoading ? (
          <Spinner className="spinner" color="danger" />
        ) : this.props.isSignedIn ? (
          <button className="button-secondary" onClick={this.toggle}>
            Add new recipe
          </button>
        ) : (
          <p>create an account to add a new recipe</p>
        )}
        <Modal isOpen={isCreateRecipeOpen} toggle={this.toggle}>
          <AddRecipeModal
            toggle={this.toggle}
            navToRecipe={this.navToRecipe}
          ></AddRecipeModal>
        </Modal>
        <div className="recipe-card-grid">
          {recipes.map((recipe) => {
            return (
              <Card>
                <CardImg
                  top
                  width="100%"
                  src={recipe.picture}
                  alt="Card image cap"
                />
                <CardBody>
                  <CardTitle>{recipe.name}</CardTitle>
                  <CardSubtitle>{recipe.cookTime}</CardSubtitle>
                  <CardText>{recipe.description}</CardText>
                  <Link to={`/recipe/${recipe._id}`}>
                    <button className="button-secondary">View Recipe</button>
                  </Link>
                </CardBody>
              </Card>
            );
          })}
        </div>
      </div>
    );
  }
}

export default MainPage;
