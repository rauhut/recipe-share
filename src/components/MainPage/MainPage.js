import React from "react";
import { Modal, Spinner } from "reactstrap";
import AddRecipeModal from "../AddRecipeModal/AddRecipeModal";
import RecipeDisplay from "../RecipeDisplay/RecipeDisplay";
import "./MainPage.css";

class MainPage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      recipes: [],
      isCreateRecipeOpen: false,
      isLoading: true,
    };
  }

  async componentDidMount() {
    try {
      const res = await fetch(
        "https://recipe-share-backend.herokuapp.com/recipes",
        {
          method: "get",
          headers: {
            "Content-Type": "application/json",
          },
        }
      );
      if (!res.ok) {
        throw new Error("Something went wrong when trying to get recipes");
      }
      const recipes = await res.json();
      this.setState({ recipes: recipes.data });
      this.setState({ isLoading: false });
    } catch (err) {
      console.log(err);
    }
  }

  toggle = () => {
    this.setState({ isCreateRecipeOpen: !this.state.isCreateRecipeOpen });
  };

  navToRecipe = (id) => {
    this.props.history.push(`/recipe/${id}`);
  };

  render() {
    const { recipes, isCreateRecipeOpen } = this.state;
    return (
      <div>
        <h1>Recipe Share</h1>
        {this.state.isLoading ? (
          <Spinner className="spinner" color="danger" />
        ) : this.props.isSignedIn ? (
          <button
            id="add-recipe"
            className="button-secondary"
            onClick={this.toggle}
          >
            Add new recipe
          </button>
        ) : (
          <h6>Create an account to add a new recipe</h6>
        )}
        <Modal isOpen={isCreateRecipeOpen} toggle={this.toggle}>
          <AddRecipeModal
            toggle={this.toggle}
            navToRecipe={this.navToRecipe}
          ></AddRecipeModal>
        </Modal>
        <RecipeDisplay recipes={recipes} />
      </div>
    );
  }
}

export default MainPage;
