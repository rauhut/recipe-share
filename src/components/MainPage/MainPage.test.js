import React from "react";
import { shallow, mount } from "enzyme";
import MainPage from "./MainPage";
import { MemoryRouter } from "react-router-dom";

describe("RecipeDisplay", () => {
  let wrapper;

  const mockPropsSignedIn = {
    isSignedIn: true,
    user: {
      id: "123456789",
      username: "testuser",
      recipes: "",
    },
  };

  const mockPropsSignedOut = {
    isSignedIn: false,
    user: {},
  };

  afterEach(() => {
    wrapper.unmount();
  });

  it("renders loading state without crashing", () => {
    wrapper = shallow(<MainPage {...mockPropsSignedIn} />);
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("renders page as logged out user without crashing", () => {
    wrapper = shallow(<MainPage {...mockPropsSignedOut} />);
    wrapper.setState({ isLoading: false });
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("renders page as logged in user without crashing", () => {
    wrapper = shallow(<MainPage {...mockPropsSignedIn} />);
    wrapper.setState({ isLoading: false });
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("can toggle the add recipe modal", () => {
    wrapper = shallow(<MainPage {...mockPropsSignedIn} />);
    wrapper.setState({ isLoading: false });
    wrapper.find('[id="add-recipe"]').simulate("click");
    expect(wrapper.state("isCreateRecipeOpen")).toEqual(true);
  });

  it("displays the recipes and hide loading spinner on api call success", (done) => {
    global.fetch = jest.fn();

    wrapper = shallow(<MainPage {...mockPropsSignedIn} />, {
      disableLifecycleMethods: true,
    });

    const spyDidMount = jest.spyOn(MainPage.prototype, "componentDidMount");

    fetch.mockImplementation(() => {
      return Promise.resolve({
        ok: true,
        json: () => {
          return Promise.resolve({
            data: [
              {
                name: "cookies",
                ingredients: ["flour", "sugar", "butter", "salt", "chocolate"],
                steps: ["step 1", "step 2", "step 3"],
                cookTime: "30 mins",
                description: "just basic cookies",
              },
              {
                name: "meatballs",
                ingredients: [
                  "ground beef",
                  "eggs",
                  "bread crumbs",
                  "garlic",
                  "onion",
                ],
                steps: ["step 1", "step 2", "step 3"],
                cookTime: "30 mins",
                description: "just basic meatballs",
              },
            ],
          });
        },
      });
    });

    const didMount = wrapper.instance().componentDidMount();

    expect(spyDidMount).toHaveBeenCalled();

    didMount.then(() => {
      wrapper.update();
      expect(wrapper.state("recipes")).toEqual([
        {
          name: "cookies",
          ingredients: ["flour", "sugar", "butter", "salt", "chocolate"],
          steps: ["step 1", "step 2", "step 3"],
          cookTime: "30 mins",
          description: "just basic cookies",
        },
        {
          name: "meatballs",
          ingredients: [
            "ground beef",
            "eggs",
            "bread crumbs",
            "garlic",
            "onion",
          ],
          steps: ["step 1", "step 2", "step 3"],
          cookTime: "30 mins",
          description: "just basic meatballs",
        },
      ]);
      expect(wrapper.state("isLoading")).toEqual(false);
      spyDidMount.mockRestore();
      fetch.mockClear();
      done();
    });
  });
});
