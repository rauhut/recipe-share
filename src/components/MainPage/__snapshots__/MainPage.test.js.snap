// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeDisplay renders loading state without crashing 1`] = `
"<div>
  <h1>
    Recipe Share
  </h1>
  <Spinner className=\\"spinner\\" color=\\"danger\\" tag=\\"div\\" type=\\"border\\">
    Loading...
  </Spinner>
  <Modal isOpen={false} toggle={[Function (anonymous)]} autoFocus={true} centered={false} scrollable={false} role=\\"dialog\\" backdrop={true} keyboard={true} zIndex={1050} fade={true} onOpened={[Function: noop]} onClosed={[Function: noop]} modalTransition={{...}} backdropTransition={{...}} unmountOnClose={true} returnFocusAfterClose={true} container=\\"body\\">
    <AddRecipeModal toggle={[Function (anonymous)]} navToRecipe={[Function (anonymous)]} />
  </Modal>
  <RecipeDisplay recipes={{...}} />
</div>"
`;

exports[`RecipeDisplay renders page as logged in user without crashing 1`] = `
"<div>
  <h1>
    Recipe Share
  </h1>
  <button id=\\"add-recipe\\" className=\\"button-secondary\\" onClick={[Function (anonymous)]}>
    Add new recipe
  </button>
  <Modal isOpen={false} toggle={[Function (anonymous)]} autoFocus={true} centered={false} scrollable={false} role=\\"dialog\\" backdrop={true} keyboard={true} zIndex={1050} fade={true} onOpened={[Function: noop]} onClosed={[Function: noop]} modalTransition={{...}} backdropTransition={{...}} unmountOnClose={true} returnFocusAfterClose={true} container=\\"body\\">
    <AddRecipeModal toggle={[Function (anonymous)]} navToRecipe={[Function (anonymous)]} />
  </Modal>
  <RecipeDisplay recipes={{...}} />
</div>"
`;

exports[`RecipeDisplay renders page as logged out user without crashing 1`] = `
"<div>
  <h1>
    Recipe Share
  </h1>
  <h6>
    Create an account to add a new recipe
  </h6>
  <Modal isOpen={false} toggle={[Function (anonymous)]} autoFocus={true} centered={false} scrollable={false} role=\\"dialog\\" backdrop={true} keyboard={true} zIndex={1050} fade={true} onOpened={[Function: noop]} onClosed={[Function: noop]} modalTransition={{...}} backdropTransition={{...}} unmountOnClose={true} returnFocusAfterClose={true} container=\\"body\\">
    <AddRecipeModal toggle={[Function (anonymous)]} navToRecipe={[Function (anonymous)]} />
  </Modal>
  <RecipeDisplay recipes={{...}} />
</div>"
`;
